### ISSUE TITLE
Util to run tests on a user instance and get the status

### ISSUE THREAD

**Issue by nikochiko** [pipalacademy/rajdhani#4]
- Should clone a repository
- Run tests with the test suite (pytest?)
- Format of status to be figured out - but the status of each task: passed/failed should be included

**Comment by nikochiko** [comment]
## Instructor experience

I think a good approach would be one where trainers are given some structure about where to put tests for each task, and they can change the information about tasks (their order, title, description) without touching code.

Here's what I'm proposing:
```
train-ride/
|__ ...
|__ some directories and files
|__ tasks.yml
|__ tests/
|__ |__ test_autocomplete.py
|__ |__ test_hello_world.py
|__ |__ test_show_trains.py
```

```yaml
# tasks.yml

tasks:
    - title: Enable home page
      description: |
           Edit `config.py` and set the `enable_home_page` to True to see the home page.
      test_module: test_hello_world

    - title: Fix the auto complete 
      description: |
           Edit `train_ride/app.py` and implement `get_stations` function.
      test_module: test_autocomplete

    - title: Show all the trains between selected stations
      description: |
           After clicking on `Search`, we should be able to see all the trains between the from and to stations.

           **Note:** Have you taken the day of the week the train runs and class into consideration?
      test_module: test_show_trains
```

Module name for tests will be fetched from the `classname` attribute in JUnit xml output of pytest. https://docs.pytest.org/en/7.1.x/how-to/output.html#creating-junitxml-format-files We can run a simple `startswith` on it and match it with the modules defined for tasks.

Advantage of this is that order, title, and description can be changed without going into the code.

## Format for serialization of task status

After the tests are run, the output can be converted to JUnit xml, and then our parser can segregate results for each `task.test_module` as defined in `tasks.yml`.
Output of each task should have:
1. The names of each test
2. Outcome of each test (passed, failed). clarification: failed includes failed/errored.
3. Logs for tests that failed

The test runner should return the results in the following format:
```
{
    "tasks": [
        {
            "title": "Enable home page",
            "description": "....",
            "tests": [
                {
                     "name": "test_home_page_enabled_is_true",
                     "outcome": "passed"
                }
            ]
        },
        {
            "title": "Fix autocomplete",
            "description": "....",
            "tests": [
                {
                     "name": "test_autocomplete_when_input_string_is_empty",
                     "outcome": "passed"
                },
                {
                     "name": "test_autocomplete_when_input_string_is_ba",
                     "outcome": "failed",
                     "logs": "Traceback ():\nAssertionError: ..."
                }
            ]
        }
    ]
}
```

This can be converted to a progress object later on.

Open questions:
1. Should we hide test results for task number 6 if task number 1 fails? (for example)
2. Would it better to use some kind of `task_id` in the return format? That way users can change the title/description without having to re-run the tests.

**Comment by anandology** [comment]
I realized that pytest may not be appropriate here an parsing what tests have passed etc. would be tricky. I've settled on writing each task as as function with the task title and description mentioned in the docstring and the function would verify if the task is done. Right now it is just checks if the task is complete or not, but it should be possible to show detailed hints when the task is not complete or not as expected.

Also, moved the tasks to new repository [rajdhani-challenge](/pipalacademy/rajdhani-challenge) to make it easier to manage.
